working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+  * Function types don't encode necessary separation between `@shared` and non-`@shared` functions.
+  * Methods like `List<T>.forEach` pose challenge because they should be usable in both `@shared` and non-`@shared` contexts.
+
+This makes us think that language changes required to achieve sound compile time delineation between `@shared` and isolate worlds are too complicated to be worth it. 
+
+###  Upgrading `dart:ffi`
+
+Introduction of _shared isolate_ allows to finally address the problem of native code invoking Dart callbacks from arbitrary threads. [`NativeCallable`](https://api.dart.dev/dev/3.3.0-246.0.dev/dart-ffi/NativeCallable-class.html) can be extended with the corresponding constructor:
+
+```dart 
+class NativeCallable<T extends Function> {
+  /// Constructs a [NativeCallable] that can be invoked from any thread.
+  ///
+  /// When the native code invokes the function [nativeFunction], the corresponding
+  /// [callback] will be synchronously executed on the same thread within a 
+  /// shared isolate corresponding to the current isolate group.
+  external factory NativeCallable.shared(
+    @DartRepresentationOf("T") Function callback,
+    {Object? exceptionalReturn});
+}
+```
+
+Invoking `NativeCallable.shared(...).nativeFunction` does not require exclusive access to a specific isolate - so it will not introduce any busy waiting. It also has a clear semantics with respect to global state: `@shared` global state is accessible and independent from the current thread and non-`@shared` state will throw an error when accessed.  
+
+## Core Library Changes
+
+### Upgrading `dart:async` capabilities
+
+#### `Zone` as the executor
+
+Consider the following code:
+
+```dart
+Future<void> foo() async {
+  await something();
+  print(1);
+}
+
+void main() async {
+  await Isolate.run(() async {
+    await foo();
+  }, shared: true);
+}
+```
+
+What happens when `Future` completes in a shared isolate? Who drives event loop of that isolate? Which thread will callbacks run on?
+
+We propose to tie answers to these questions to the existing concept: `Zone`. Root `Zone` of each isolate will responsible for managing event loop of that isolate. A `Zone` might have a thread pool associated with it or it might be driven externally by the embedder. 
+
+Consider for example the code like `fut.then((v) { /* ... */ })`. When `Future` `fut` completes the invocation of the callback is going to be managed by the `Zone` to which it is bound: if `Zone` has thread pool associated with it and we are not on one of the threads associated with that pool then a task will be posted to the pool instead of executing callback on the current thread. This makes it easy to understand and control how asynchronous execution moves between threads. 
+
+If you get a hold of `Zone` from another isolate (`Zone` should probably be `Shareable`) then you should be able to inject a callback into the event loop owned by that `Zone`.  
```

Comment:
The root zone might be shareable, other zones are unlikely to be, since they contain zone functions which have no reason to require being shareable. Or zone variables which are also unrestricted. (Zones exist on platforms with no isolates, the zone functions can do anything.)